### PR TITLE
feat(tui): query extended underline support using DECRQSS

### LINKED ISSUE #26744: Implementing undercurl support in windows terminal
### Problem

Windows did not support underline syles other than the regular one. no undercurl or double line . but with the latest merge one month ago these has been implemented and planned for  https://github.com/microsoft/terminal/pull/16097.

in the latest canary you can see this in action using powershell >= 6  (Cmd as far as i know does not support escape chars. correct me if i am wrong )

using both 
```
Write-Host "`e[4:3mtest`e[0m"
Write-Host "`e[4:3;58:2::000:255:000mtest`e[0m"
```
will print both colored non colored undercurl 

image

but neovim in windows won't show these even when modifying the t_Ce and t_Cs variables

```
if vim.fn.has("termguicolors") == 1 then
    vim.o.termguicolors = true
    -- Enable undercurl
    vim.api.nvim_set_var('t_Cs', '\\e[4:3m')
    vim.api.nvim_set_var('t_Ce', '\\e[4:0m')
end
```

Can this be implemented early so when the final release is done this wil be ready?
the latest canary also supports all other styles of underline like dotted and double underlines.

### Expected behavior

to show undercurls and other styles of  underlines

### PR DESCRIPTION
**Pull request by zeertzjq**
Close #26744
Close #27629

### CODE REVIEW

**Comment by gpanders** [comment]
> A problem with this approach is that it may interfere with truecolor
> support detection, as both use SGR followed by DECRQSS, and there is no
> synchronization because they are sent by two different processes.

Perhaps we should move the truecolor query into `tui/input.c` as well and merge the two.

**Comment by zeertzjq** [comment]
Yes that's a possible solution, and will also allow truecolor detection to work with `--remote-ui`.

File: src/nvim/tui/input.c [hunk 1]
```diff
@@ -553,6 +553,10 @@ static void handle_term_response(TermInput *input, const TermKeyKey *key)
   if (termkey_interpret_string(input->tk, key, &str) == TERMKEY_RES_KEY) {
     assert(str != NULL);
 
+    if (key->type == TERMKEY_TYPE_DCS && strnequal(str, S_LEN("1$r4:3"))) {
+      tui_enable_extended_underline(input->tui_data);
+    }
```
[the remark this answers is not in the document]
Comment (zeertzjq) [reply src/nvim/tui/input.c]: Thanks

File: src/nvim/tui/tui.c [hunk 2]
```diff
@@ -225,6 +225,25 @@ void tui_handle_term_mode(TUIData *tui, TermMode mode, TermModeState state)
   }
 }
 
+/// Query the terminal emulator to see if it supports extended underline.
+///
+/// Try to set an undercurl, followed by a DECRQSS query.
+static void tui_query_extended_underline(TUIData *tui)
+{
+  out(tui, S_LEN("\x1b[0m\x1b[4:3m\x1bP$qm\x1b\\"));
```
Comment (tusharsnx) [inline src/nvim/tui/tui.c:233]: nit: 

Add a comment on why we are resetting the attribute first so it's not lost with time.

**Comment by zeertzjq** [comment]
> > A problem with this approach is that it may interfere with truecolor
> > support detection, as both use SGR followed by DECRQSS, and there is no
> > synchronization because they are sent by two different processes.
> 
> Perhaps we should move the truecolor query into `tui/input.c` as well and merge the two.

Actually it may be fine as long as the autocommand isn't deleted prematurely.

**Pull request merged**